### PR TITLE
dataset symlinks provided

### PR DESCRIPTION
**Pull request by pawel-big-lebowski**
Signed-off-by: Pawel Leszczynski <leszczynski.pawel@gmail.com>

### Problem

A `SymlinkDatasetFacet` has been introduced in spec recently (https://github.com/OpenLineage/OpenLineage/pull/936) and it allows a dataset to be identified by multiple `(name, namespace)` tuples. We need to modify Marquez to handle it, as currently many joins to dataset table are done directly based on `name` and `namespace` value. 

Part of: #2066

### Solution

This PR contains:
* a refactor of current Marquez database model to allow alternative dataset names by creating an extra `dataset_symlinks` table with dataset name. 
* removal of dataset's name from dataset table.
* implementation of `SymlinkDatasetFacet` logic. 

> **Note:** All database schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

### Checklist

- [x] You've [signed-off](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#sign-your-work) your work
- [ ] Your changes are accompanied by tests (_if relevant_)
- [ ] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] You've updated the [`CHANGELOG.md`](https://github.com/MarquezProject/marquez/blob/main/CHANGELOG.md#unreleased) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned your `.sql` database schema migration according to [Flyway's naming convention](https://flywaydb.org/documentation/concepts/migrations#naming) (_if relevant_)
- [ ] You've included a [header](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#copyright--license) in any source code files (_if relevant_)

### CODE REVIEW

**Comment by mobuchowski** [comment]
>removal of dataset's name from dataset table.

Why?

**Comment by mobuchowski** [comment]
@pawel-big-lebowski let me try to summarize changes to model: after this change, internally `datasets` table don't represent "full" dataset, since name is dropped - getting `dataset identifier` requires joining with symlinks table. `datasets` table still contains `namespace_name` - it's the same namespace name that is used by `primary` symlink.

Primary symlink is the one that is being send in regular dataset name, rather the ones in `SymlinksDatasetFacet`. That suggest question: what happens if two jobs produce the same dataset, and they have the identifiers in different order; that is one has namespace X and name A as primary, and second one has namespace Y and name B? Does the first one "win" and dataset is forever known as (X, A) pair?

**Comment by pawel-big-lebowski** [comment]
> @pawel-big-lebowski let me try to summarize changes to model: after this change, internally `datasets` table don't represent "full" dataset, since name is dropped - getting `dataset identifier` requires joining with symlinks table. `datasets` table still contains `namespace_name` - it's the same namespace name that is used by `primary` symlink.
> 

Yes, you're right. The reason behind droping a `name` column  is to make sure that no-one now or later tries to retrieve dataset by filtering `name` column in `dataset` table. Whenever this is required, `dataset_symlinks` table should be joined to make sure we look for alternative names. 

The `namespace_name` column existed before together with `namespace_uuid` field. In my opinion, it should be removed but I found out it's out of scope for this PR. 

> Primary symlink is the one that is being send in regular dataset name, rather the ones in `SymlinksDatasetFacet`. That suggest question: what happens if two jobs produce the same dataset, and they have the identifiers in different order; that is one has namespace X and name A as primary, and second one has namespace Y and name B? Does the first one "win" and dataset is forever known as (X, A) pair?

Yes, the first (namespace/name) becomes primary and I think it's the best we can do.

**Review by julienledem — COMMENTED** [review]
Overall this looks good to me. Is it going to conflict with the delete feature @mobuchowski is working on?

**Review by collado-mike — COMMENTED** [review]
Given the problem the dataset symlinking aims to solve is to resolve lineage for jobs that may be accessing datasets by different names, I'm wondering if we can scope the symlinks functionality to _just_ the lineage API. As is, its impact is pervasive and, as we saw with the job symlinks, that can have unintended consequences. I think we can start by scoping this to just the lineage API, then adjust other APIs as needed.

**Comment by pawel-big-lebowski** [comment]
> Given the problem the dataset symlinking aims to solve is to resolve lineage for jobs that may be accessing datasets by different names, **I'm wondering if we can scope the symlinks functionality to _just_ the lineage API.** As is, its impact is pervasive and, as we saw with the job symlinks, that can have unintended consequences. I think we can start by scoping this to just the lineage API, then adjust other APIs as needed.

In order to confirm that, one should know all the Marquez features to make sure datasets are not joined by name whenever symlinks should be applied. This creates an assumption which is not stated explicitly, so is error prone in future. That's why I am in favour of an assumption that `dataset_symlinks` should be always use to retrieve a dataset by name. If a dataset has two different names, it gets returned to matter how we request it. 

Performance wise, this should not affect postgresql. It's just an extra join of index column to retrieve extra column (not used for filtering) and an uncorrelated subquery to get possible symlinks run before existing queries. Postgres is smart and joins on foreign keys, which are just applied on result rows, should be fast. Full-scans digging jsonb columns are slow, we know that and there is proposal to fix it.

To sum up, there is a tradeoff "(1) not clean assumption in DB modelling" vs "(2) DB performance risk that should not happen but cannot be confirmed due to lack of perf tests". One one hand, I would go with (2) as it gives greater chance to make it clean in future. On the other hand, I don't mind going with (1) as I haven't worked that much with Marquez DB perf issues.

File: api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+CREATE TABLE dataset_symlinks (
+  dataset_uuid      UUID,
```
Comment (pawel-big-lebowski) [inline api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql:4]: TODO: index on `dataset_uuid` missing

**Comment by mobuchowski** [comment]
@julienledem great point. I talked with @pawel-big-lebowski and we aligned to use `datasets_view` to provide joined data from `datasets` and `datasets_symlinks`. This way, most of the SQL busywork behind this change will be opaque to queries.

With that, only issue is that "namespaces" aren't unique anymore - methods like 
```
List<Dataset> findAll(String namespaceName, int limit, int offset)
```
still need to filter by `is_primary` field to not return duplicates.

File: api/src/main/java/marquez/db/Columns.java [hunk 2]
```diff
@@ -80,6 +80,9 @@ private Columns() {}
   public static final String FIELD_UUIDS = "field_uuids";
   public static final String LIFECYCLE_STATE = "lifecycle_state";
 
+  /* DATASET SYMLINK ROW COLUMNS */
+  public static final String IS_PRIMARY = "is_primary";
```
Comment (wslulciuc) [inline api/src/main/java/marquez/db/Columns.java:84]: Is the `is_primary` column used to indicated that it's the _primary_ name of the dataset?

File: api/src/main/java/marquez/db/DatasetDao.java [hunk 3]
```diff
@@ -42,9 +43,14 @@
 @RegisterRowMapper(DatasetMapper.class)
 public interface DatasetDao extends BaseDao {
   @SqlQuery(
-      "SELECT EXISTS ("
+      "WITH symlink AS (\n"
```
Comment (wslulciuc) [inline api/src/main/java/marquez/db/DatasetDao.java:46]: _Minor_: `symlink` -> `symlinks`

File: api/src/main/java/marquez/db/Columns.java (hunk 2 above)
Comment (wslulciuc) [reply api/src/main/java/marquez/db/Columns.java:84]: Ahh it totally is 👍

File: api/src/main/java/marquez/db/DatasetSymlinkDao.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+public interface DatasetSymlinkDao extends BaseDao {
+
+  default DatasetSymlinkRow upsertDatasetSymlinkRow(
+      UUID uuid, String name, UUID namespaceUuid, boolean isPrimary, Instant now) {
+    doUpsertDatasetSymlinkRow(uuid, name, namespaceUuid, isPrimary, now);
+    return findDatasetSymlinkByNamespaceUidAndName(namespaceUuid, name).orElseThrow();
+  }
+
+  @SqlQuery("SELECT * FROM dataset_symlinks WHERE namespace_uuid = :namespaceUuid and name = :name")
+  Optional<DatasetSymlinkRow> findDatasetSymlinkByNamespaceUidAndName(
```
Comment (wslulciuc) [inline api/src/main/java/marquez/db/DatasetSymlinkDao.java:27]: _Typo_: `findDatasetSymlinkByNamespaceUidAndName()` -> `findDatasetSymlinkByNamespaceUuidAndName()`

File: api/src/main/java/marquez/db/DatasetSymlinkDao.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      UUID namespaceUuid, String name);
+
+  @SqlUpdate(
+      "INSERT INTO dataset_symlinks ("
```
Comment (wslulciuc) [inline api/src/main/java/marquez/db/DatasetSymlinkDao.java:31]: In [`V46__dataset_symlinks.sql`](https://github.com/MarquezProject/marquez/blob/feature/dataset-symlinks/api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql) a `type` column is defined, but never set here. Do we need the column?

File: api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  name              VARCHAR(255) NOT NULL,
+  namespace_uuid    UUID REFERENCES namespaces(uuid),
+  type              VARCHAR(64),
```
Comment (wslulciuc) [inline api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql:7]: Do we need `type` defined in `dataset_symlinks`? Or, would it be helpful to know the type of symlink?

File: api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  created_at        TIMESTAMP NOT NULL,
+  updated_at        TIMESTAMP NOT NULL,
+  UNIQUE (namespace_uuid, name)
+);
+
+INSERT INTO dataset_symlinks (dataset_uuid, name, namespace_uuid, is_primary, created_at, updated_at)
+SELECT d.uuid, d.name, d.namespace_uuid, TRUE, d.created_at, d.updated_at FROM datasets d;
+
+ALTER TABLE datasets DROP CONSTRAINT datasets_namespace_uuid_name_key;
+ALTER TABLE datasets DROP COLUMN name;
```
Comment (wslulciuc) [inline api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql:18]: Mind if we split these changes into two separate migrations?

* `V46__add_dataset_symlinks.sql`
* `V46.1__drop_name_column_in_datasets.sql`

Comment (wslulciuc) [reply api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql:18]: Do we need to drop the `name` column from `datasets`? I'm wondering, can `datasets.name` be viewed as the primary name? Or, will the `name` column in `datasets_view` (introduced in https://github.com/MarquezProject/marquez/pull/2032) be considered the dataset name (=_primary_)?

File: api/src/main/java/marquez/db/DatasetSymlinkDao.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.jdbi.v3.sqlobject.statement.SqlUpdate;
+
+@RegisterRowMapper(DatasetSymlinksRowMapper.class)
+public interface DatasetSymlinkDao extends BaseDao {
+
+  default DatasetSymlinkRow upsertDatasetSymlinkRow(
+      UUID uuid, String name, UUID namespaceUuid, boolean isPrimary, Instant now) {
+    doUpsertDatasetSymlinkRow(uuid, name, namespaceUuid, isPrimary, now);
+    return findDatasetSymlinkByNamespaceUidAndName(namespaceUuid, name).orElseThrow();
```
[the remark this answers is not in the document]
Comment (wslulciuc) [reply api/src/main/java/marquez/db/DatasetSymlinkDao.java:23]: `RETURNING` is great, but only if the row as changed. To get around this, most of our queries update the `updated_at` column even though no changes have occurred; which, to me, isn't ideal and something I'd like to move away from.

**Comment by wslulciuc** [comment]
> To sum up, there is a tradeoff "(1) not clean assumption in DB modelling" vs "(2) DB performance risk that should not happen but cannot be confirmed due to lack of perf tests". On one hand, I would go with (2) as it gives great chance to make it clean in future. On the other hand, I don't mind going with (1) as I haven't worked that much with Marquez DB perf issues.

@pawel-big-lebowski I think @collado-mike's comment: "_As is, its impact is pervasive and, as we saw with the job symlinks, that can have unintended consequences_" wasn't necessarily performance related (though there was follow up work on job symlinks needed to improve performance and implementation), but rather unexpected behavior given the number of queries modified. 

We're hoping to add [load testing](https://github.com/MarquezProject/marquez/blob/main/docs/load-testing.md) to CI at some point, see #2047

**Review by wslulciuc — CHANGES_REQUESTED** [review]
I would also update our [`CHANGELOG.md`](https://github.com/MarquezProject/marquez/blob/main/CHANGELOG.md) with this sweet sweet feature!

**Comment by wslulciuc** [comment]
> The namespace_name column existed before together with namespace_uuid field. In my opinion, it should be removed but I found out it's out of scope for this PR.

@pawel-big-lebowski mind opening an issue to capture this change as follow up work?

File: api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql (hunk 1 above)
Comment (wslulciuc) [reply api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql:4]: Don't forget to add this index ;)

File: api/src/main/java/marquez/db/DatasetSymlinkDao.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/DatasetSymlinkDao.java]: I saw that pattern in other pieces of Marquez and decided to reuse it.

File: api/src/main/java/marquez/db/DatasetDao.java (hunk 3 above)
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/DatasetDao.java]: code got removed when switching to `dataset_view`

File: api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql (hunk 7 above)
Comment (pawel-big-lebowski) [reply api/src/main/resources/marquez/db/migration/V46__dataset_symlinks.sql]: Based on the feedback, `name` column won't be dropped to mitigate the risk of migration issues and limit unexpected behavior changes.

File: api/src/main/java/marquez/db/DatasetSymlinkDao.java (hunk 5 above)
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/DatasetSymlinkDao.java:31]: This column is part of a spec to contain some description of an identifier for example: `table` may be a type for symlink names containing table name.

I forgot to make use of the column. Thanks for spotting this. 

I added some code to save a `type` and included it in unit test `marquez.db.OpenLineageDaoTest#testUpdateMarquezModelDatasetWithSymlinks` to make sure the type from OL event is written to Marquez DB.

**Comment by pawel-big-lebowski** [comment]
@collado-mike , @pawel-big-lebowski  and @wslulciuc, thank you for great reviews and feedback.

Changes applied according to that: 
 * will reuse a  `dataset_view` introduced by Maciej,
 * won't drop `name` column in `datasets` table to avoid risky migrations and limit amount of possible unexpected behavior changes.

File: api/src/main/java/marquez/db/DatasetSymlinkDao.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+          + "updated_at"
+          + ") VALUES ( "
+          + ":uuid, "
+          + ":name, "
+          + ":namespaceUuid, "
+          + ":isPrimary, "
+          + ":type, "
+          + ":now, "
+          + ":now) "
+          + "ON CONFLICT (name, namespace_uuid) DO NOTHING")
```
Comment (mobuchowski) [inline api/src/main/java/marquez/db/DatasetSymlinkDao.java:47]: If we're adding new SQL blocks, please use Java 17's text blocks feature.

File: api/src/main/resources/marquez/db/migration/V47__dataset_symlinks.sql [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  UNIQUE (namespace_uuid, name)
+);
+
+CREATE INDEX dataset_symlinks_dataset_uuid on dataset_symlinks (dataset_uuid);
+
+INSERT INTO dataset_symlinks (dataset_uuid, name, namespace_uuid, is_primary, created_at, updated_at)
+SELECT d.uuid, d.name, d.namespace_uuid, TRUE, d.created_at, d.updated_at FROM datasets d;
+
+DROP TYPE IF EXISTS DATASET_NAME;
+CREATE TYPE DATASET_NAME AS (
```
Comment (mobuchowski) [inline api/src/main/resources/marquez/db/migration/V47__dataset_symlinks.sql:20]: Why do we need it?

File: api/src/main/resources/marquez/db/migration/V47__dataset_symlinks.sql [hunk 11]
```diff
@@ -0,0 +1,24 @@
+/* SPDX-License-Identifier: Apache-2.0 */
```
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply api/src/main/resources/marquez/db/migration/V47__dataset_symlinks.sql:1]: sure, will do that once the review is over ;-)

File: api/src/main/resources/marquez/db/migration/V47__dataset_symlinks.sql (hunk 10 above)
Comment (pawel-big-lebowski) [reply api/src/main/resources/marquez/db/migration/V47__dataset_symlinks.sql:20]: because of this:
```
WHERE CAST((:namespaceName, :datasetName) AS DATASET_NAME) = ANY(d.dataset_symlinks)
```
https://github.com/MarquezProject/marquez/pull/2087/files#diff-c687ca258e023847591c9b8c044adddaae1e7c1f6f9483928bf2871050cf6a63R77

I thought it's necessary in postgresql to check if pair of varchars is contained in array of pairs of varchars.

Comment (mobuchowski) [reply api/src/main/resources/marquez/db/migration/V47__dataset_symlinks.sql:20]: Ok, so you can do `WHERE namespace_name IN X AND name in Y` because it can match namespace, and match name, but those belong to different symlinks, right?

Comment (pawel-big-lebowski) [reply api/src/main/resources/marquez/db/migration/V47__dataset_symlinks.sql:20]: > Ok, so you can do `WHERE namespace_name IN X AND name in Y` because it can match namespace, and match name, but those belong to different symlinks, right?

Exactly, @mobuchowski

File: api/src/main/java/marquez/db/DatasetSymlinkDao.java (hunk 9 above)
Comment (wslulciuc) [reply api/src/main/java/marquez/db/DatasetSymlinkDao.java:47]: +1

File: api/src/main/java/marquez/db/DatasetDao.java [hunk 12]
```diff
@@ -310,6 +312,10 @@ default Dataset upsertDatasetMeta(
         createNamespaceDao()
             .upsertNamespaceRow(
                 UUID.randomUUID(), now, namespaceName.getValue(), DEFAULT_NAMESPACE_OWNER);
+    DatasetSymlinkRow symlinkRow =
+        createDatasetSymlinkDao()
+            .upsertDatasetSymlinkRow(
+                UUID.randomUUID(), datasetName.getValue(), namespaceRow.getUuid(), true, "", now);
```
Comment (wslulciuc) [inline api/src/main/java/marquez/db/DatasetDao.java:318]: Can we default the symlink type to `UNKNOWN` or `null` if we don't know the symlink type rather than an empty string?

File: api/src/main/java/marquez/db/DatasetSymlinkDao.java (hunk 5 above)
Comment (wslulciuc) [reply api/src/main/java/marquez/db/DatasetSymlinkDao.java:31]: Thanks! ❤️

File: api/src/main/java/marquez/db/OpenLineageDao.java [hunk 13]
```diff
@@ -540,6 +545,35 @@ default DatasetRecord upsertLineageDataset(
             formatNamespaceName(ds.getNamespace()),
             DEFAULT_NAMESPACE_OWNER);
 
+    DatasetSymlinkRow symlink =
+        datasetSymlinkDao.upsertDatasetSymlinkRow(
+            UUID.randomUUID(),
+            formatDatasetName(ds.getName()),
+            dsNamespace.getUuid(),
+            true,
+            "",
```
Comment (wslulciuc) [inline api/src/main/java/marquez/db/OpenLineageDao.java:554]: Similar comment as above, I would avoid using an empty string for the symlink type and default to `UNKNOWN` or `null`.

File: api/src/main/java/marquez/service/models/DatasetSymlink.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+import java.time.Instant;
+import java.util.Optional;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Value;
+
+@Value
+public class DatasetSymlink {
```
Comment (wslulciuc) [inline api/src/main/java/marquez/service/models/DatasetSymlink.java:17]: Is class `DatasetSymlink` used?

**Review by wslulciuc — APPROVED** [review]
Great work, @pawel-big-lebowski! I left some additional comments, but looks good to merge overall once they are resolved 💯 🥇

File: api/src/main/java/marquez/db/DatasetDao.java [hunk 15]
```diff
@@ -42,9 +43,14 @@
 @RegisterRowMapper(DatasetMapper.class)
 public interface DatasetDao extends BaseDao {
   @SqlQuery(
-      "SELECT EXISTS ("
+      "WITH symlink AS (\n"
+          + "    SELECT dataset_uuid\n"
+          + "    FROM dataset_symlinks\n"
+          + "    JOIN namespaces ON namespaces.uuid = dataset_symlinks.namespace_uuid\n"
+          + "    WHERE namespaces.name=:namespaceName and dataset_symlinks.name=:datasetName\n"
```
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/DatasetDao.java]: i am using a view defined in Maciej's PR.

File: api/src/main/java/marquez/db/OpenLineageDao.java (hunk 13 above)
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/OpenLineageDao.java:554]: let's go with `null`

File: api/src/main/java/marquez/db/DatasetDao.java (hunk 12 above)
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/db/DatasetDao.java:318]: `null` is fine
